demos/tutorials/Gibbs.py

```diff
@@ -0,0 +1,222 @@
+# the ``Linear_RTO`` sampler for :math:`\mathbf{x}`.
+#
+# This is done in CUQIpy as follows:
+
+# Define Gibbs sampler
+sampler = Gibbs(posterior, {'x': Linear_RTO, ('d', 'l'): Conjugate})
+
+# Run sampler
+samples = sampler.sample(Ns=1000, Nb=200)
+
+# %%
+# Analyze results
+# ---------------
+#
+# After sampling we can inspect the results. The samples are stored
+# as a dictionary with the variable names as keys. Samples for each 
+# variable is stored as a CUQIpy Samples object which contains the
+# many convenience methods for diagnostics and plotting of MCMC samples.
+
+# Plot credible intervals for the signal
+samples['x'].plot_ci(exact=probinfo.exactSolution)
+
+# %%
+# Trace plot for d
+samples['d'].plot_trace(figsize=(8,2))
+
+# %%
+# Trace plot for l
+samples['l'].plot_trace(figsize=(8,2))
+
+# %%
+# Switching to a piecewise constant prior
+# ---------------------------------------
+#
+# Notice that while the sampling went well in the previous example,
+# the posterior distribution did not match the characteristics of
+# the exact solution. We can improve this result by switching to a
+# piecewise constant prior for :math:`\mathbf{x}`.
+#
+# One choice is the Laplace difference prior, which assumes a
+# Laplace distribution for the differences between neighboring
+# elements of :math:`\mathbf{x}`. That is,
+#
+# .. math::
+#
+#     \mathbf{x} \sim \text{Laplace_diff}(\mathbf{0}, d^{-1}),
+#
+# which means that :math:`x_i-x_{i-1} \sim \mathrm{Laplace}(0, d^{-1})`.
+#
+# This prior is implemented in CUQIpy as the ``Laplace_diff`` distribution.
+# To update our model we simply need to replace the ``GMRF`` distribution
+# with the ``Laplace_diff`` distribution. Note that the Laplace distribution
+# is defined via a scale parameter, so we invert the parameter :math:`d`.
+#
+# This laplace distribution and new posterior can be defined as follows:
+
+# Define new distribution for x
+x = Laplace_diff(np.zeros(n), lambda d: 1/d)
+
+# Define new joint distribution
+joint = JointDistribution([d, l, x, y])
+
+# Define new posterior by conditioning on the data
+posterior = joint(y=y_obs)
+
+print(posterior)
+
+
+# %%
+# Gibbs Sampler (with Laplace prior)
+# ----------------------------------
+#
+# Using the same approach as ealier we can define a Gibbs sampler
+# for this new hierarchical model. The only difference is that we
+# now need to use a different sampler for :math:`\mathbf{x}` because
+# the ``Linear_RTO`` sampler only works for Gaussian distributions.
+#
+# In this case we use the UnadjustedLaplaceApproximation sampler
+# for :math:`\mathbf{x}`. We also use an approximate Conjugate
+# sampler for :math:`d` which approximately samples from the
+# posterior distribution of :math:`d` conditional on the other
+# variables in an efficient manner. For more details see e.g.
+# `this paper <https://arxiv.org/abs/2104.06919>`.
+
+# Define sampling strategy
+sampling_strategy = {
```

Comment:
I like this way of having the sampling_strategy split off from the call to Gibbs. I think this makes it clearer what is going on.

Reply:
I tend to agree. Started using it in both places :)